## Natural Language Processing (NLP)
#### Machine Learning labolatorium 15.01.2026
Opracowanie: Kajetan Kowalski 421071

---

In [ ]:
import numpy as np

### 1. Utworzenie danych (listy zdań, które symulujką raporty geologiczne lub wyniki misji kosmicznych).

In [ ]:
texts = [
    "Odkryto złoża żelaza w warstwach osadowych",
    "Badania geofizyczne wskazują anomalię magnetyczną",
    "Misja kosmiczna potwierdziła obecność minerałów na Marsie",
    "Wiercenia w rejonie rud miedzi wykazały wysoką koncentrację metalu",
    "Obserwacje satelitarne ujawniły aktywność wulkaniczną na powierzchni planety Wenus",
    "Pomiary sejsmiczne ujawniły nieciągłość struktur skalnych w obrębie uskoku tektonicznego",
    "Radar penetrujący grunt wykazał obecność struktur podpowierzchniowych na asteroidzie",
    "Analiza rdzeni wiertniczych potwierdziła obecność siarczków niklu na dużej głębokości"
]

Najważniejsze słowa kluczowe:

|Zdanie|Skłowa kluczowe|
|------|---------------|
|1|złoża, żelazo, geologia, osady|
|2|badania, geogizyka, magnetyzm, anomalia|
|3|misja kosmiczna, Mars, eksploracja kosmosu, minerały|
|4|wiercenia, miedź, metal, górnictwo, złoże|
|5|obserwacje satelitarne, aktywność wulkaniczna, Wenus, geologia|
|6|pomiary sejsmiczne, sejsmologia, nieciągłość , uskok tektoniczny, tektonika, geofizyka|
|7|radar, struktury podpowierzchniowe, asteroida, badania kosmiczne, obrazowanie|
|8|rdzenie wiertnicze, siarczki niklu, geochemia, surowce metaliczne|


Informacje przydatne do klasyfikacji:

Do klasyfikacji zdań kluczowe są informacje wskazujące na obiekt badań (np. Ziemia, planeta, asteroida), zastosowaną metodę badawczą (np. wiercenia, pomiary sejsmiczne, obserwacje satelitarne) oraz rodzaj analizowanego materiału lub zjawiska (np. minerały, złoża, aktywność wulkaniczna). Zestaw tych cech pozwala jednoznacznie odróżnić zdania dotyczące geologii ziemskiej od opisów eksploracji kosmosu.

### 2. Tokenizacja

Tokenizacja to proces, w którym każde słowo w zdaniu jest przypisywane do unikalnej liczby całkowitej, aby komputer mógł je przetwarzać. W tej części zadania zamienisz swoje zdania na liczby.

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer(num_words=100)  # ograniczamy słownik do 100 najczęściej występujących słów
tokenizer.fit_on_texts(texts)          # wstaw listę tekstów
print(tokenizer.word_index)

{'na': 1, 'w': 2, 'obecność': 3, 'potwierdziła': 4, 'ujawniły': 5, 'struktur': 6, 'odkryto': 7, 'złoża': 8, 'żelaza': 9, 'warstwach': 10, 'osadowych': 11, 'badania': 12, 'geofizyczne': 13, 'wskazują': 14, 'anomalię': 15, 'magnetyczną': 16, 'misja': 17, 'kosmiczna': 18, 'minerałów': 19, 'marsie': 20, 'wiercenia': 21, 'rejonie': 22, 'rud': 23, 'miedzi': 24, 'wykazały': 25, 'wysoką': 26, 'koncentrację': 27, 'metalu': 28, 'obserwacje': 29, 'satelitarne': 30, 'aktywność': 31, 'wulkaniczną': 32, 'powierzchni': 33, 'planety': 34, 'wenus': 35, 'pomiary': 36, 'sejsmiczne': 37, 'nieciągłość': 38, 'skalnych': 39, 'obrębie': 40, 'uskoku': 41, 'tektonicznego': 42, 'radar': 43, 'penetrujący': 44, 'grunt': 45, 'wykazał': 46, 'podpowierzchniowych': 47, 'asteroidzie': 48, 'analiza': 49, 'rdzeni': 50, 'wiertniczych': 51, 'siarczków': 52, 'niklu': 53, 'dużej': 54, 'głębokości': 55}


#### 2.1 Sprawdzenie słowa z najmniejszym indeksem

In [ ]:
first_word = tokenizer.index_word[1]
count = tokenizer.word_counts[first_word]

print(f"Słowo z indeksem 1 to: '{first_word}'")
print(f"Wystąpiło ono w tekstach {count} razy.")

typ = "NAJCZĘŚCIEJ" if count == max(
    tokenizer.word_counts.values()) else "NAJRZADZIEJ"

print(f"WNIOSEK: Indeks 1 otrzymuje słowo {typ} występujące.")


Słowo z indeksem 1 to: 'na'
Wystąpiło ono w tekstach 4 razy.
WNIOSEK: Indeks 1 otrzymuje słowo NAJCZĘŚCIEJ występujące.


#### 2.2 Słowa niewystępujące w pozostałych tekstach

In [ ]:
original_texts = texts[:-4]
new_texts = texts[-4:]


def get_clean_words(text_list):
    words_set = set()
    for text in text_list:
        clean_text = text.lower().replace('.', '').replace(',', '')
        words_set.update(clean_text.split())
    return words_set

original_words = get_clean_words(original_texts)
new_words = get_clean_words(new_texts)

unique = new_words - original_words

print(f"Słowa, których nie było wcześniej: {unique}")
print(f"Jest ich: {len(unique)}")

Słowa, których nie było wcześniej: {'sejsmiczne', 'satelitarne', 'planety', 'radar', 'asteroidzie', 'pomiary', 'wulkaniczną', 'niklu', 'obrębie', 'obserwacje', 'penetrujący', 'uskoku', 'wenus', 'podpowierzchniowych', 'ujawniły', 'tektonicznego', 'głębokości', 'skalnych', 'dużej', 'wykazał', 'powierzchni', 'nieciągłość', 'aktywność', 'siarczków', 'grunt', 'analiza', 'rdzeni', 'wiertniczych', 'struktur'}
Jest ich: 29


### 3. Zamiana tekstuy na sekwencje liczbowych tokenów

In [ ]:
sequences = tokenizer.texts_to_sequences(texts)  # wstaw listę tekstów
print("Wynik sequences:\n", sequences)

Wynik sequences:
 [[7, 8, 9, 2, 10, 11], [12, 13, 14, 15, 16], [17, 18, 4, 3, 19, 1, 20], [21, 2, 22, 23, 24, 25, 26, 27, 28], [29, 30, 5, 31, 32, 1, 33, 34, 35], [36, 37, 5, 38, 6, 39, 2, 40, 41, 42], [43, 44, 45, 46, 3, 6, 47, 1, 48], [49, 50, 51, 4, 3, 52, 53, 1, 54, 55]]


In [ ]:
for i, seq in enumerate(sequences):
    is_integers = all(isinstance(x, int) for x in seq)
    all_in_vocab = all(x in tokenizer.index_word for x in seq)
    print(f"Zdanie {i+1}:")
    print(f"  - Długość: {len(seq)} tokenów")
    print(f"  - Czy same liczby całkowite? {is_integers}")
    print(f"  - Czy wszystkie tokeny w słowniku? {all_in_vocab}")
    print(f"  - Oryginał: '{texts[i]}'")

print()
print("DEKODOWANIE PIERWSZEGO ZDANIA:")
first_seq = sequences[0]
print(f"Sekwencja liczb: {first_seq}")


for num in first_seq:
    word = tokenizer.index_word.get(num, "NIEZNANE")
    check_index = tokenizer.word_index[word]

    print(f"Liczba {num} --> Słowo '{word}' (Sprawdzenie w word_index: {check_index})")

Zdanie 1:
  - Długość: 6 tokenów
  - Czy same liczby całkowite? True
  - Czy wszystkie tokeny w słowniku? True
  - Oryginał: 'Odkryto złoża żelaza w warstwach osadowych'
Zdanie 2:
  - Długość: 5 tokenów
  - Czy same liczby całkowite? True
  - Czy wszystkie tokeny w słowniku? True
  - Oryginał: 'Badania geofizyczne wskazują anomalię magnetyczną'
Zdanie 3:
  - Długość: 7 tokenów
  - Czy same liczby całkowite? True
  - Czy wszystkie tokeny w słowniku? True
  - Oryginał: 'Misja kosmiczna potwierdziła obecność minerałów na Marsie'
Zdanie 4:
  - Długość: 9 tokenów
  - Czy same liczby całkowite? True
  - Czy wszystkie tokeny w słowniku? True
  - Oryginał: 'Wiercenia w rejonie rud miedzi wykazały wysoką koncentrację metalu'
Zdanie 5:
  - Długość: 9 tokenów
  - Czy same liczby całkowite? True
  - Czy wszystkie tokeny w słowniku? True
  - Oryginał: 'Obserwacje satelitarne ujawniły aktywność wulkaniczną na powierzchni planety Wenus'
Zdanie 6:
  - Długość: 10 tokenów
  - Czy same liczby całkowite?

Na podstawie przeprowadzonej analizy procesu tokenizacji tekstu można sformułować następujące wnioski:
* Każde zdanie po przetworzeniu przez tokenizer zostało poprawnie zamienione na sekwencję liczb całkowitych. Potwierdza to, że mechanizm texts_to_sequences() mapuje słowa na jednoznaczne identyfikatory liczbowe, zgodnie z definicją słownika tokenizera.

* Wszystkie wygenerowane tokeny posiadają odpowiadające im słowa w słowniku tokenizera (word_index), co zostało potwierdzone wartością `True` dla każdego zdania. Oznacza to, że w analizowanym zbiorze danych nie występują tokeny spoza słownika oraz że tokenizacja została wykonana na zbiorze znanych słów.

* Długość sekwencji tokenów jest bezpośrednio zależna od liczby słów w zdaniu źródłowym - krótsze zdania generują krótsze sekwencje, natomiast dłuższe zdania prowadzą do powstania sekwencji o większej liczbie tokenów. Potwierdza to zachowanie tokenizera polegające na przypisywaniu jednego tokenu do jednego słowa.

* Dekodowanie pierwszego zdania wykazało jednoznaczną i spójną relację pomiędzy tokenami liczbowymi a odpowiadającymi im słowami. Każdy numer tokenu został poprawnie odwzorowany na właściwe słowo, a jego ponowne sprawdzenie w word_index potwierdziło zgodność mapowania w obu kierunkach (słowo <=> indeks).

* Uzyskane wyniki potwierdzają poprawność działania tokenizera oraz spójność wewnętrznej reprezentacji tekstu, co stanowi podstawę do dalszego etapu przetwarzania danych tekstowych, takiego jak uzupełnianie sekwencji (padding) lub wykorzystanie ich jako wejścia do modeli sieci neuronowych.

### 4. Padding dla modeli DL
Modele deep learningowe (=neuronowe) wymagają, aby wszystkie sekwencje wejściowe miały taką samą długość. Dlatego stosujemy padding, czyli dopasowujemy krótsze sekwencje do długości najdłuższej (lub ustalonej).

In [ ]:
from tensorflow.keras.preprocessing.sequence import pad_sequences
padded = pad_sequences(sequences, padding='post')
print(padded)


print(f"Liczba zdań (wierszy): {padded.shape[0]}")
print(f"Długość każdego zdania (kolumny): {padded.shape[1]}")

[[ 7  8  9  2 10 11  0  0  0  0]
 [12 13 14 15 16  0  0  0  0  0]
 [17 18  4  3 19  1 20  0  0  0]
 [21  2 22 23 24 25 26 27 28  0]
 [29 30  5 31 32  1 33 34 35  0]
 [36 37  5 38  6 39  2 40 41 42]
 [43 44 45 46  3  6 47  1 48  0]
 [49 50 51  4  3 52 53  1 54 55]]
Liczba zdań (wierszy): 8
Długość każdego zdania (kolumny): 10


- Jak zmieniła się długość sekwencji po paddingu?
  - W wyniku operacji `pad_sequences` wszystkie sekwencje wejściowe zostały ujednolicone do wspólnej długości wynoszącej 10 tokenów. Długość ta została narzucona przez najdłuższe zdanie w zbiorze danych, więc wszystkie krótsze zdania zostały wydłużone, aby pasowały do tego wymiaru.

- Jakie liczby reprezentują brakujące miejsca (padding)?
  - Brakujące miejsca w krótszych sekwencjach zostały wypełnione liczbą 0. Zera te pełnią funkcję wypełniacza i nie niosą ze sobą wartości semantycznej (nie odpowiadają żadnemu słowu ze słownika). Dzięki zastosowaniu parametru `padding='post'`, zera zostały dodane na końcu sekwencji, zachowując naturalną kolejność słów na początku wektora.

- Dlaczego takie uzupełnienie jest potrzebne w modelach sieci neuronowych?
  - Standaryzacja długości wektora wejściowego jest fundamentalnym wymogiem dla modeli sieci neuronowych. Biblioteki takie jak TensorFlow/Keras operują na tensorach (wielowymiarowych macierzach). Aby utworzyć tensor z grupy zdań, każdy wiersz macierzy musi mieć identyczną liczbę kolumn. Stały wymiar danych umożliwia przetwarzanie wielu próbek jednocześnie, co jest kluczowe dla wydajności obliczeniowej na procesorach GPU. Bez paddingu niemożliwe byłoby wykonanie operacji mnożenia wag sieci przez dane wejściowe.

### 5. Przygotowanie zbioru danych do klasyfikacji.

In [ ]:
texts = [
    "Odkryto złoża żelaza w warstwach osadowych",                                  # 0 (Geo)
    "Badania geofizyczne wskazują anomalię magnetyczną",                           # 0 (Geo)
    "Misja kosmiczna potwierdziła obecność minerałów na Marsie",                   # 1 (Kosmos)
    "Wiercenia w rejonie rud miedzi wykazały wysoką koncentrację metalu",          # 0 (Geo)
    "Obserwacje satelitarne ujawniły aktywność wulkaniczną na powierzchni planety Wenus", # 1 (Kosmos)
    "Pomiary sejsmiczne ujawniły nieciągłość struktur skalnych w obrębie uskoku tektonicznego", # 0 (Geo)
    "Radar penetrujący grunt wykazał obecność struktur podpowierzchniowych na asteroidzie",     # 1 (Kosmos)
    "Analiza rdzeni wiertniczych potwierdziła obecność siarczków niklu na dużej głębokości",     # 0 (Geo)
    "Analiza próbek skał wykazała podwyższoną zawartość pierwiastków śladowych",   # 0 (Geo)
    "Profil geologiczny ujawnił obecność głębokich struktur krystalicznych",      # 0 (Geo)
    "Sonda orbitalna zarejestrowała zmiany składu atmosfery planety",              # 1 (Kosmos)
    "Teleskop kosmiczny wykrył ślady lodu wodnego na powierzchni księżyca"         # 1 (Kosmos)
]

labels = [0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1]

tokenizer = Tokenizer(num_words=100)
tokenizer.fit_on_texts(texts)
sequences = tokenizer.texts_to_sequences(texts)
padded = pad_sequences(sequences, padding='post')
labels = np.array(labels)

max_length = padded.shape[1]

print(f"Liczba zdań: {len(texts)}")
print(f"Długość wejścia (max_length): {max_length}")

print("Dane do treningu:")
print(f"Tekst: {texts[4]}")
print(f"Sekwencja: {padded[4]}")
print(f"Etykieta: {labels[4]}")

Liczba zdań: 12
Długość wejścia (max_length): 10
Dane do treningu:
Tekst: Obserwacje satelitarne ujawniły aktywność wulkaniczną na powierzchni planety Wenus
Sekwencja: [32 33  6 34 35  1  7  8 36  0]
Etykieta: 1


### 6. Prosty model w Keras

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Flatten, Dense

model = Sequential([
    Embedding(input_dim=100, output_dim=8, input_length=max_length),
    Flatten(),
    Dense(1, activation='sigmoid')  # dla klasyfikacji binarnej
])

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

model.fit(padded, labels, epochs=10)

Epoch 1/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step - accuracy: 0.6667 - loss: 0.6915
Epoch 2/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - accuracy: 0.6667 - loss: 0.6873
Epoch 3/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - accuracy: 0.6667 - loss: 0.6831
Epoch 4/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - accuracy: 0.8333 - loss: 0.6789
Epoch 5/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.9167 - loss: 0.6747
Epoch 6/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.9167 - loss: 0.6705
Epoch 7/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.9167 - loss: 0.6663
Epoch 8/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step - accuracy: 0.9167 - loss: 0.6621
Epoch 9/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - accuracy: 0.9167 - loss: 0.6579
Epoch 10/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - accuracy: 0.9167 - loss: 0.6537


- Jak embeddingi reprezentują słowa w przestrzeni liczbowej?

  - Embeddingi reprezentują każde słowo jako wektor liczb rzeczywistych o stałym wymiarze. Wektory te są uczone podczas treningu modelu, a nie narzucane z góry. Słowa występujące w podobnych kontekstach otrzymują zbliżone reprezentacje wektorowe, co pozwala modelowi uchwycić relacje semantyczne pomiędzy pojęciami, np. terminami geologicznymi lub kosmicznymi.

- Czy model poprawnie odróżnia raporty geologiczne od kosmicznych?
  - Na podstawie wyników treningu można stwierdzić, że model uczy się rozróżniać obie klasy. Dokładność (accuracy) wzrosła z 0.683 do 0.917, a funkcja straty systematycznie maleje, co wskazuje na poprawne dopasowanie do danych treningowych. Ze względu na bardzo mały zbiór danych, wynik ten ma charakter demonstracyjny, jednak potwierdza, że model potrafi wychwycić różnice pomiędzy raportami geologicznymi a kosmicznymi, choć jeżeli temat dotyczy obu zagadnień może się on mylić.

Ze względu na ograniczoną liczbę przykładów, uzyskane wyniki nie pozwalają na ocenę zdolności generalizacji modelu, lecz potwierdzają poprawność zastosowanego podejścia.

### 7. Test na nowym zdaniu

In [ ]:
test_text = ["Nowe badania wskazują złoża niklu na asteroidzie"]

seq = tokenizer.texts_to_sequences(test_text)

padded_test = pad_sequences(
    seq,
    maxlen=max_length,
    padding='post'
)

prediction = model.predict(padded_test)
print(prediction)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 74ms/step
[[0.48704708]]


In [ ]:
score = prediction[0][0]

print("-" * 30)
print(f"Zdanie: '{test_text[0]}'")
print(f"prawdopodobienstwo: {score:.4f}")

if score < 0.5:
    print("Werdykt modelu: GEOLOGIA (klasa 0)")
    print("(Model uznał, że słowa 'złoża/nikiel' są ważniejsze)")
else:
    print("Werdykt modelu: KOSMOS (klasa 1)")
    print("(Model uznał, że słowo 'asteroida' jest ważniejsze)")

------------------------------
Zdanie: 'Nowe badania wskazują złoża niklu na asteroidzie'
prawdopodobienstwo: 0.4870
Werdykt modelu: GEOLOGIA (klasa 0)
(Model uznał, że słowa 'złoża/nikiel' są ważniejsze)


- Czy model prawidłowo sklasyfikował zdanie?
  - Model nie sklasyfikował zdania prawidłowo -zdanie dotyczy asteroidy, czyli obiektu kosmicznego, dlatego poprawną klasą powinna być klasa 1 (Kosmos). Model przypisał jednak zdanie do klasy geologia, ponieważ w procesie uczenia większą wagę nadał słowom takim jak `złoża` i `niklu`, które w zbiorze treningowym częściej występowały w kontekście geologicznym. Wartość predykcji (0.4843) znajduje się blisko progu decyzyjnego, co wskazuje na niepewność modelu.
- Jak można poprawić działanie modelu przy nowych słowach?
  - Działanie modelu można poprawić poprzez zwiększenie różnorodności i liczebności danych uczących, w szczególności poprzez dodanie zdań łączących terminologię geologiczną z kontekstem kosmicznym (np. surowce na asteroidach). Dodatkowo pomocne byłoby zastosowanie mechanizmu obsługi słów spoza słownika, użycie wstępnie wytrenowanych embeddingów lub zastąpienie warstwy Flatten warstwą rekurencyjną (LSTM lub GRU), co pozwoliłoby lepiej uchwycić kontekst całego zdania.

### 8. Pipeline

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import FunctionTransformer

def tokenize_pad(X):
    seq = tokenizer.texts_to_sequences(X)
    return pad_sequences(seq, maxlen=max_length, padding='post')

pipeline = Pipeline([
    ('tokenize_pad', FunctionTransformer(tokenize_pad, validate=False)),
    ('model', model)
])

Jak pipeline chroni przed błędami przy nowych słowach?

- Pipeline chroni nie tyle przed samymi "nowymi słowami" (bo to robi Tokenizer), co przed błędami proceduralnymi (niespójnością danych). Najczęstszym błędem jest próba wrzucenia do modelu zdania o innej długości niż ta, na której model był uczony. Pipeline blokuje parametr max_length. Niezależnie od tego, czy zostanie wpisane jedno słowo, czy dłuższy tekst, funkcja tokenize_pad wewnątrz pipeline'u zawsze przytnie lub dopełni (pad) dane do idealnego wymiaru, którego oczekuje model. Użytkownik końcowy nie musi wiedzieć, co to jest tokenizacja czy padding - wrzuca surowy tekst i otrzymuje wynik. Eliminuje to ryzyko, że człowiek zapomni bądź nie posiada wiedzy o jakimś z kroków przetwarzania wstępnego.

### 9. Test pipeline

In [ ]:
new_texts = ["Eksploracja Marsa wykazała obecność żelaza i niklu"]
raw_prediction = pipeline.predict(new_texts)
score = raw_prediction[0][0]

print("-" * 30)
print(f"Zdanie: '{new_texts[0]}'")
print(f"prawdopodobienstwo: {score:.4f}")

if score > 0.5:
    print("KLASYFIKACJA: Kosmos / Misje (1)")
    print("Słowa o kosmosie zdominowały słowa dotyczące minerałów i geologii")
else:
    print("KLASYFIKACJA: Geologia / Złoża (0)")
    print("Słowa o minerałach zdominowały kontekst zdania.")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step
------------------------------
Zdanie: 'Eksploracja Marsa wykazała obecność żelaza i niklu'
prawdopodobienstwo: 0.4934
KLASYFIKACJA: Geologia / Złoża (0)
Słowa o minerałach zdominowały kontekst zdania.


- Czy pipeline poprawnie przetworzył dane i dokonał predykcji?
  - Tak, pipeline poprawnie przetworzył dane i wykonał predykcję. Surowy tekst został automatycznie poddany tokenizacji oraz paddingowi, a następnie przekazany do modelu Keras bez konieczności ręcznego wywoływania tych kroków. Oznacza to, że pipeline działa zgodnie z założeniem i umożliwia bezpośrednią predykcję na podstawie nieprzetworzonego tekstu wejściowego. Innym aspektem jest poprawność, predykcji - wystąpił adekwatny problem jak omawiany w punkcie 8 - niepoprawny wynik.
- Jak można rozbudować pipeline o preprocessing, np. usuwanie stop-words lub stemming?
  - Pipeline można rozbudować poprzez dodanie dodatkowych etapów przetwarzania tekstu przed tokenizacją. Przykładowo, możliwe jest wprowadzenie kroku usuwania słów nieistotnych, normalizacji tekstu (np. sprowadzenie do małych liter) lub stemmingu bądź lematyzacji, co pozwoliłoby ograniczyć warianty tej samej formy wyrazu. Takie operacje zwiększają spójność danych wejściowych i mogą poprawić skuteczność modelu, szczególnie przy większych i bardziej zróżnicowanych zbiorach danych.